internal/storagenode/logstream/sync.go

```diff
@@ -367,18 +367,32 @@ func (lse *Executor) SyncInit(_ context.Context, srcReplica varlogpb.LogStreamRe
 		)
 	}
 
-	syncRange = snpb.SyncRange{
-		FirstLLSN: uncommittedLLSNBegin,
-		LastLLSN:  srcRange.LastLLSN,
-	}
-
 	// NOTE: When the replica has all log entries, it returns its range of logs and non-error results.
 	// In this case, this replica remains executorStateSealing.
 	// Breaking change: previously it returns ErrExist when the replica has all log entries to replicate.
-	if lastCommittedLLSN == syncRange.LastLLSN {
+	if lastCommittedLLSN == srcRange.LastLLSN {
 		return snpb.SyncRange{FirstLLSN: types.InvalidLLSN, LastLLSN: types.InvalidLLSN}, nil
-		//err = fmt.Errorf("log stream: sync init: already enough logs: %w", verrors.ErrExist)
-		//return
+	}
+
+	if uncommittedLLSNBegin < srcRange.FirstLLSN { // The source replica may have been trimmed.
+		lastCommittedLLSN = srcRange.FirstLLSN - 1
+		uncommittedLLSNBegin = lastCommittedLLSN + 1
+		// NOTE: The version and high watermark are not correct.
+		lse.lsc.storeReportCommitBase(types.InvalidVersion, types.InvalidGLSN, uncommittedLLSNBegin) // new uncommittedLLSNBegin
+		// NOTE: Since prefix logs has been trimmed, the local low and
+		// high watermarks are invalid. It will be reset by a commit
+		// triggered by SyncReplicate.
+		// It means that metadata of replica in the status of
+		// `LEARNING` is not credible because learning replica, which
+		// is the destination of synchronization, has invalid low and
+		// high watermarks.
+		lse.lsc.setLocalLowWatermark(varlogpb.InvalidLogEntryMeta())
```

Comment:
https://github.com/kakao/varlog/blob/c4d0c8d5926e834ffd6f5a7def2649bbed2a7075/internal/storagenode/logstream/executor.go#L539-L543

Because the new lowWatermark is larger than the stored latest log entry, it is unlikely that the previous log will be trimmed. Is it intended to be trimmed late?

Reply:
Yes, it is. However, I think it will be a problem when recovering the log stream. I will fix it in this PR.

Reply:
> Yes, it is. However, I think it will be a problem when recovering the log stream. I will fix it in this PR.

@hungryjang 
Unlike the above comment, I will fix it in follow-up PR because it needs too many changes.

For more explanation:
The hole can be a problem while the storage node reloads the log stream. To solve it, we can either remove the prefix logs trimmed or handle the hole while recovering. As you know, we need to improve the performance of recovering a log stream. So it can be easier to clean the logs up when processing SyncInit.
Another problem with cleaning the logs during handling SyncInit is that the current Trim API of the storage layer only supports GLSN, but SyncRange is only represented by LLSN. We should fix it up later.

Reply:
@ijsong Okay